8306914: Implement JEP 458: Launch Multi-File Source-Code Programs

sormuras opened the pull request:
Please review this change set implementing [JEP 458](https://openjdk.org/jeps/458) "Launch Multi-File Source-Code Programs" by introducing a new entry-point `SourceLauncher` in package `com.sun.tools.javac.launcher` of module `jdk.compiler`. The code of the new entry-point and other new classes introduced by this pull request is based on the original implementation (now removed `Main.java` file). It extends it to dynamically resolve required types "on-the-fly" using an in-memory class loader capable to compile Java source files - applying a strict name-based pattern as described in the JEP.

### OpenJDK's Demo Programs

OpenJDK's demo programs can be found in the [src/demo/share/jfc](https://github.com/openjdk/jdk/tree/master/src/demo/share/jfc) directory. These multi-file source-code programs can be launched via `java ${PROGRAM}` once this implementation is integrated; with `${PROGRAM}` being one of the following files:

- [x] `src/demo/share/jfc/CodePointIM/CodePointIM.java`
- [x] `src/demo/share/jfc/FileChooserDemo/FileChooserDemo.java`
- [x] `src/demo/share/jfc/Font2DTest/Font2DTest.java`
- [x] `src/demo/share/jfc/J2Ddemo/java2d/J2Ddemo.java`
- [x] `src/demo/share/jfc/Metalworks/Metalworks.java`
- [x] `src/demo/share/jfc/Notepad/Notepad.java`
- [x] `src/demo/share/jfc/SampleTree/SampleTree.java`
- [x] `src/demo/share/jfc/SwingSet2/SwingSet2.java`
- [x] `src/demo/share/jfc/TableExample/TableExample3.java`
- [x] `src/demo/share/jfc/TableExample/TableExample4.java`
- [x] `src/demo/share/jfc/TransparentRuler/transparentruler/Ruler.java`

The following demo programs do not start out of the box:

- `src/demo/share/jfc/TableExample/TableExample.java` — requires a database driver on the class path
- `src/demo/share/jfc/TableExample/TableExample2.java` —  requires a database driver on the class path
- `src/demo/share/jfc/Stylepad/Stylepad.java` — requires `src/demo/share/jfc/Notepad/Notepad.java`

Tested tier1-tier5.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8306915](https://bugs.openjdk.org/browse/JDK-8306915) to be approved

### Issues
 * [JDK-8306914](https://bugs.openjdk.org/browse/JDK-8306914): Implement JEP 458: Launch Multi-File Source-Code Programs (**Enhancement** - P4)
 * [JDK-8306915](https://bugs.openjdk.org/browse/JDK-8306915): Implementation of JEP Launch Multi-File Source-Code Programs (**CSR**)


### Reviewers
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**)
 * [Jim Laskey](https://openjdk.org/census#jlaskey) (@JimLaskey - **Reviewer**) ⚠️ Review applies to [63551153](https://git.openjdk.org/jdk/pull/13712/files/63551153002f9f191d707b3ec038435a93d3018f)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13712/head:pull/13712` \
`$ git checkout pull/13712`

Update a local copy of the PR: \
`$ git checkout pull/13712` \
`$ git pull https://git.openjdk.org/jdk.git pull/13712/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13712`

View PR using the GUI difftool: \
`$ git pr show -t 13712`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13712.diff">https://git.openjdk.org/jdk/pull/13712.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13712#issuecomment-1763989666)

jddarcy commented:
A few high-level comments after an initial read over the PR:

I recommend declared the new top-level cases final (or sealed) where possible.

I suggest leaving some "breadcrumbs" in other option-handling locations in the code base to provide maintenance reminders that if a "relevant" option is added, it should be added to the RelevantJavacOptions list.

sormuras commented:
/open

lahodaj commented on src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/MemoryContext.java:
```diff
@@ -0,0 +1,194 @@
+    }
+
+    byte[] compileJavaFileByName(String name) throws Fault {
+        var firstDollarSign = name.indexOf('$'); // [package . ] name [ $ enclosed ]
```
Nit - in principle, `$` is a valid identifier character:
```
jshell> Character.isJavaIdentifierStart('$')
$1 ==> true
````

So, in principle, the user's code may have classes with `$` in their names. OTOH, this seems like a very rare case given the context.

lahodaj commented on src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/MemoryFileManager.java:
```diff
@@ -0,0 +1,90 @@
+ * risk.  This code and its internal interfaces are subject to change
+ * or deletion without notice.</b></p>
+ */
+final class MemoryFileManager extends ForwardingJavaFileManager<JavaFileManager> {
```
Just FWIW, I note this class does not support reading from the classfiles, so everything will be load from the source files during compilation.

lahodaj commented on src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/ProgramDescriptor.java:
```diff
@@ -0,0 +1,119 @@
+                  .map(Path::toString)
+                  .filter(string -> !string.isEmpty())
+                  .map(string -> string.replace('/', '.'))
+                  .map(string -> string.replace('\\', '.'))
```
Nit: I'd be a bit careful with the `\` -> `.` replacement. `\` is a valid file name character on UNIX-like system. Better use `File.separatorChar` or `sourceRootPath.getFileSystem().getSeparator()`. Although the practical impact is probably very low.

sormuras replied:
Addressed this in https://github.com/openjdk/jdk/pull/13712/commits/093ab1cef7383323deb74bffa511b81109ca4dd6

lahodaj commented on src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/SourceLauncher.java:
```diff
@@ -0,0 +1,268 @@
+    public static void main(String... args) throws Throwable {
+        try {
+            new SourceLauncher(System.err)
+                    .checkSecurityManager()
```
Nit, fwiw: the SecurityManager check is not completely safe (as the user's program could install SM while providing -Djava.security.manager=allow), and then further compilation may be invoked, failing on SM. OTOH, given the state of SM, I don't think this needs changes.

lahodaj commented on test/langtools/tools/javac/launcher/Run.java:
```diff
@@ -0,0 +1,63 @@
+
+        PrintStream prev = System.out;
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        try (PrintStream out = new PrintStream(baos, true)) {
```
So, I am looking at this, and I am not sure how it can work and test what is needed. If the compilation fails during on-demand class loading, we stop using `System.exit`, and it seems to me this code cannot handle that, and so that there's probably no test testing that scenario?

I mean:

```
---Test.java
public class Test {
    public static void main(String... args) throws Throwable {
        Class.forName("Extra");
    }
}
----Extra.java
public class Broken
---

sormuras replied:
There is a test in `MultiFileSourceLauncherTests` that checks this with a broken second file: `testSecondUnitWithSyntaxError`, including an assertion for the syntax error being emitted.

And I recall to have seen the error code `2` in a manual test - but there's no automated test for this.

lahodaj reviewed (commented):
I did an initial pass through the code. Added some small comments, mostly for future consideration, and one slightly bigger one around testing case where on-demand compilation fails.

AlanBateman commented on src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/MemoryContext.java:
```diff
@@ -0,0 +1,257 @@
+        var modulePathModules = modulePathFinder.findAll().stream().map(ModuleReference::descriptor).map(ModuleDescriptor::name).toList();
+        if (!modulePathModules.isEmpty()) {
+            var modulePathConfiguration = bootLayer.configuration().resolveAndBind(modulePathFinder, ModuleFinder.of(), Set.copyOf(modulePathModules));
+            var modulePathLayer = ModuleLayer.defineModulesWithOneLoader(modulePathConfiguration, List.of(bootLayer), parent).layer();
```
Since you don't need the Controller then it may be simpler to use just `bootLayer.defineModulesWithOneLoader(modulePathConfiguration, parent);`

sormuras replied:
Will change the line as suggested.

AlanBateman commented on src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/MemoryContext.java:
```diff
@@ -0,0 +1,257 @@
+        var mainClassNamePackageName = mainClassName.substring(0, lastDotInMainClassName);
+        memoryController.addOpens(module, mainClassNamePackageName, getClass().getModule());
+
+        return memoryLayer.findLoader(applicationModule.name()).loadClass(mainClassName);
```
The defining class loader for the main class will be memoryClassLoader, right?

AlanBateman commented on src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/MemoryContext.java:
```diff
@@ -0,0 +1,257 @@
+        var memoryFinder = new MemoryModuleFinder(inMemoryClasses, applicationModule, descriptor);
+        var memoryConfig = parentLayer.configuration().resolveAndBind(memoryFinder, ModuleFinder.of(), Set.of(applicationModule.name()));
+        var memoryClassLoader = new MemoryClassLoader(inMemoryClasses, parentLoader, applicationModule, descriptor, this::compileJavaFileByName);
+        var memoryController = ModuleLayer.defineModules(memoryConfig, List.of(parentLayer), __ -> memoryClassLoader);
```
It would be helpful if the comment were expanded to explain why memoryClassLoader is needed here. ClassLoaders created by the defineModulesXXX methods are capable to loading modules in memory file systems, the issue is the dynamic compilation (I think).

sormuras replied:
I'll add a comment explaining that the `MemoryClassLoader` is required here to compile referenced types from their source files on-the-fly.

sormuras replied:
Comments added, Alan.

Archie, that's true for the "in-memory" part - but this class loader also tries to find and compile new classes from the default file system on-the-fly, which the class loader in JShell does not (need to do). If both do share enough code, a consolidation would be good to have.

AlanBateman commented on src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/MemoryModuleFinder.java:
```diff
@@ -0,0 +1,149 @@
+import java.util.Set;
+import java.util.stream.Stream;
+
+import jdk.internal.module.Resources;
```
This is related to robustness/security so probably okay to have javac use it.

sormuras commented on src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/MemoryContext.java:
```diff
@@ -0,0 +1,257 @@
+        var modulePathFinder = createModuleFinderFromModulePath();
+        var modulePathModules = modulePathFinder.findAll().stream().map(ModuleReference::descriptor).map(ModuleDescriptor::name).toList();
+        if (!modulePathModules.isEmpty()) {
+            var modulePathConfiguration = bootLayer.configuration().resolveAndBind(modulePathFinder, ModuleFinder.of(), Set.copyOf(modulePathModules));
```
Will change the line as suggested.

turbanoff commented on src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/SourceLauncher.java:
```diff
@@ -0,0 +1,301 @@
+                Class<?> nextClass;
+                try {
+                    nextClass = Class.forName(name, true, firstClass.getClassLoader());
+                }  catch (ClassNotFoundException ignore) {
```
```suggestion
                } catch (ClassNotFoundException ignore) {
```

turbanoff commented on src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/SourceLauncher.java:
```diff
@@ -0,0 +1,294 @@
+            Class<?> actualClass;
+            try {
+                actualClass = Class.forName(actualName, false, firstClass.getClassLoader());
+            }  catch (ClassNotFoundException ignore) {
```
```suggestion
            } catch (ClassNotFoundException ignore) {
```

sormuras commented:
Thanks for pointing out the extra space characters, @turbanoff -- I will remove them and some more in a clean-up commit.

sormuras commented:
@JimLaskey Thanks for the review - addressed all comments via https://github.com/openjdk/jdk/pull/13712/commits/70540efef8d1487ac3a2522e7299716ea3f46bc1

lahodaj commented on test/jdk/java/util/stream/GathererTest.java:
```diff
@@ -349,7 +349,7 @@ public void testCompositionOfStatefulGatherers(Config config) {
     @MethodSource("configurations")
     public void testMassivelyComposedGatherers(Config config) {
-        final int ITERATIONS = 512; // Total number of compositions is 1 + (iterations*2)
+        final int ITERATIONS = 128; // Total number of compositions is 1 + (iterations*2)
```
Sorry, but I think it is not appropriate to do a change like this here. It should have its own PR and discussion.

sormuras replied:
Yes, I will revert this change and let @viktorklang-ora fine-tune it in another PR.

AlanBateman replied:
Viktor has the change in JDK-8321124 to dial it down. Right now it requires too much stack on x86_32 systems used in the GHA testing.

lahodaj commented on test/langtools/tools/javac/launcher/src/p/q/CLTest.java:
```diff
@@ -1,10 +1,12 @@
 /*
- * Copyright (c) 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
```
Probably:
```suggestion
 * Copyright (c) 2018, 2023, Oracle and/or its affiliates. All rights reserved.
```
(not sure, but the file existed before, it is just moved, AFAIK.)

sormuras replied:
Yes. It was moved from `.../launcher/src` to `.../launcher/p/q`, thus committing as suggested.

lahodaj reviewed (commented):
Overall seems reasonable to me, with two comments inline - one minor, but the other one is the change to the Gatherers tests - sorry, but I don't think it is appropriate to bury that change in this PR/commit.

lahodaj reviewed (approved):
Looks reasonable to me.

sormuras commented:
/integrate
